### PR TITLE
perf - Reduce TSO waker churn

### PR DESCRIPTION
**Pull request by mingley**
## Summary

This PR hardens TSO sender/response coordination in `src/pd/timestamp.rs` to remove a lock-contention lost-wake risk while preserving throughput optimizations.

Quality objective:
- eliminate possible sender stalls under lock contention,
- keep wake/waker overhead low on hot paths,
- preserve timestamp correctness and backpressure behavior.

### Why this change exists
The prior optimization path used a custom lock-contention handshake (`try_lock` + shared waiting flag + custom wake coordination). That reduced contention overhead but still left an interleaving where sender progress could depend on a future external wake, creating stall risk.

This change removes that fragile handshake and relies on the async mutex waiter path for lock-contention wakeup guarantees.

### Deep Dive: Concurrency Model
There are two coordination loops:
- Sender loop: `TsoRequestStream::poll_next`
- Response loop: `handle_tso_responses`

Shared state:
- `pending_requests: Mutex<VecDeque<RequestGroup>>`
- `sending_future_waker: AtomicWaker`

Updated behavior:
1. Lock contention in sender path now uses `pending_requests.lock().poll(cx)`.
This places the task on the mutex wait queue and gives deterministic wakeup when the lock becomes available.

2. Response path wake policy remains conditional:
- compute `was_full` before `allocate_timestamps`,
- wake sender only on `full -> non-full` transition.

3. No lock-contention side-channel flag is used anymore.
This removes a race-prone protocol and narrows wake logic to queue-capacity transitions.

### Why this does not break backward compatibility
- No public API/type/signature changes.
- Request/response protobuf semantics unchanged.
- Timestamp allocation logic unchanged (`allocate_timestamps` invariants are preserved).
- Backpressure contract unchanged (`MAX_PENDING_COUNT` still bounds in-flight groups).
- Background task error/termination semantics remain explicit and unchanged from current branch behavior.

Net effect: internal scheduling becomes safer under contention without changing externally observable client API behavior.

## Benchmark Results

Measured with a temporary standalone Rust microbenchmark harness (`/tmp/pr529_tso_bench`) to compare policy-level overhead for pre-PR logic (before) and latest PR logic (after). Harness was run in `--release` mode and not committed to this repository.

Benchmark configuration:
- 40 measured samples (8 warmup samples)
- 30,000,000 iterations per sample
- Full-queue transition period: 1 in 1024 iterations
- Measurement: ns/op (median, p05, p95)

### Response Wake Policy

| Variant | Median ns/op | p05 ns/op | p95 ns/op |
|---|---:|---:|---:|
| before (always wake) | 1.726378 | 1.553182 | 1.897283 |
| after (conditional wake) | 0.243229 | 0.237556 | 0.258467 |

Speedup (median): **7.10x**

### Self-Waker Registration Policy

| Variant | Median ns/op | p05 ns/op | p95 ns/op |
|---|---:|---:|---:|
| before (always register) | 2.422287 | 2.358501 | 2.468619 |
| after (register only when full) | 0.255310 | 0.234999 | 0.266172 |

Speedup (median): **9.49x**

Notes:
- These numbers isolate internal wake/waker policy overhead; they are not end-to-end PD/TSO RPC latency.
- Results validate directional impact of hot-path scheduling changes.

## File Scope

- `src/pd/timestamp.rs`
- `Cargo.toml` (branch-level context change already present in PR)

## Testing Done

Executed locally:

1. `cargo +1.93.0 test pd::timestamp --lib` -> pass (`14 passed`)
2. `cargo +1.93.0 test` -> pass (`67 unit passed`, `49 doc passed`)
3. Benchmark harness rerun (`cargo +1.93.0 run --release` in `/tmp/pr529_tso_bench`) -> metrics above

Focused concurrency coverage in `src/pd/timestamp.rs` includes:
- `poll_next_waits_on_mutex_when_lock_is_contended`
- `poll_next_registers_self_waker_when_pending_queue_is_full`
- `poll_next_does_not_register_self_waker_when_queue_not_full`
- `handle_tso_responses_wakes_sender_when_queue_transitions_from_full`
- `handle_tso_responses_does_not_wake_sender_when_queue_was_not_full`
- `handle_tso_responses_wakes_sender_once_for_each_full_to_non_full_transition`
- timestamp allocation invariant and error-path tests

## Compatibility

No public API surface change is introduced by this PR.

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive unit tests for timestamp request handling, queue behavior, and edge cases.

* **Chores**
  * Optimized async task management and response processing for improved reliability.

### CODE REVIEW

**Comment by mingley** [comment]
Addressed the CodeRabbit feedback in commit `cd6aa8f`:

1. Background TSO task errors are now surfaced via logs
- Updated `src/pd/timestamp.rs` so the spawned task wraps `run_tso(...)` and logs any `Err(...)` with context (`warn!`) instead of silently discarding it.

2. Benchmark assumptions are now explicit
- Added comments in `benches/tso_waker_policy.rs` for `FULL_EVERY` / `FULL_WINDOW`.
- Added inline notes clarifying old vs new wake behavior (unconditional vs transition-only wake).

3. Benchmark doc now includes interpretation/staleness guidance
- Updated `doc/tso_waker_criterion.md` to state these numbers are point-in-time and should be re-run after changes to `src/pd/timestamp.rs` or the benchmark.
- Added a note that the response benchmark speedup reflects amortized skipped wake calls under this simulation pattern.

**Comment by mingley** [comment]
@coderabbitai review

**Comment by mingley** [comment]
Addressed local review finding #1 in `bba9a1c`:
- `src/pd/timestamp.rs`: changed stream loop from `while let Some(Ok(resp))` to `while let Some(resp)` + `let resp = resp?;` so gRPC stream errors are propagated out of `run_tso` and logged by the spawned-task wrapper instead of being silently treated as normal termination.

**Comment by mingley** [comment]
@coderabbitai review\n\nPlease run a fresh review on the latest head (). Focus on lock-contention wake safety and the new response-loop wake tests.

**Comment by mingley** [comment]
@coderabbitai review\n\nPlease run a fresh review on the latest head: 10f685c0d23f24b8aa94244c480344a750b9ddcb. Focus on lock-contention wake safety and the new response-loop wake tests.

File: src/pd/timestamp.rs [hunk 1]
```diff
@@ -57,8 +58,13 @@ impl TimestampOracle {
...
 
-        // Start a background thread to handle TSO requests and responses
-        tokio::spawn(run_tso(cluster_id, pd_client, request_rx));
+        // Start a background task to handle TSO requests and responses.
+        // If it exits with an error, log it explicitly so root cause is preserved.
+        tokio::spawn(async move {
+            if let Err(err) = run_tso(cluster_id, pd_client, request_rx).await {
+                warn!("TSO background task exited with error: {:?}", err);
+            }
+        });
```
[the remark this answers is not in the document]
Comment (mingley) [reply src/pd/timestamp.rs:69]: Addressed in 7530dbf. `TimestampOracle::new` now uses `tokio::runtime::Handle::try_current()` and returns an internal error when no runtime is present, instead of unconditionally calling `tokio::spawn`.

File: src/pd/timestamp.rs [hunk 2]
```diff
@@ -216,3 +245,411 @@ fn allocate_timestamps(
...
+        let polled = stream.as_mut().poll_next(&mut cx);
+        assert!(matches!(polled, Poll::Pending));
+        assert_eq!(wake_counter.wakes.load(Ordering::SeqCst), 0);
+
+        // Releasing the lock should wake the pending mutex waiter.
+        drop(lock_guard);
+
+        let polled = stream.as_mut().poll_next(&mut cx);
+        assert!(matches!(polled, Poll::Ready(Some(_))));
+    }
```
[the remark this answers is not in the document]
Comment (mingley) [reply src/pd/timestamp.rs:503]: Addressed in 7530dbf. `TsoRequestStream` now persists a `lock_owned()` future across polls (`pending_requests_lock`) so mutex contention wakeups are retained. The lock-contention test now asserts automatic wake on lock release before repolling.

**Comment by mingley** [comment]
@coderabbitai please re-review the latest updates in 7530dbf; both major findings are addressed and threads are resolved.
